File: SAR/debug.py
import cProfile
import Search
import FeatureSimilarity
import pickle
from fuzzywuzzy import process
import pstats
from pstats import SortKey
from matplotlib import pyplot as plt
import numpy as np
import os
import gc

def main(songValue, pathList):
     # Find most similar song using cosine similarity
    numOfSongs = 5
    similarSongs = Search.multiProcessing(
        Search.findSimilarSongs, 32, songValue, pathList, numOfSongs)

    # Sort the list by the similarity score
    sortedSimilarSongs = sorted(
        similarSongs, key=Search.takeSecond, reverse=True)
    if len(sortedSimilarSongs) > numOfSongs:
        sortedSimilarSongs = sortedSimilarSongs[:numOfSongs]

    print("Similar Songs: ")
    for i in range(len(sortedSimilarSongs)):
        print("Found Song: " + sortedSimilarSongs[i][0].split('\0')
              [0]+" by " + sortedSimilarSongs[i][0].split('\0')[1])
        print("Cosine Similarity: " + str(sortedSimilarSongs[i][1]))

def preformanceTest():
    # Ask for the song title
    songTitle = input("Enter the song title: ")

    # Find the song in the database
    songKey = Search.fuzzyGetSongTitle(songTitle, os.path.abspath(os.getcwd()) + r'\tmp\namelist.pickle', 40)

    if len(songKey) > 1:
        print("Multiple songs found:")
        for j in range(len(songKey)):
            print(str(j+1)+". "+songKey[j][0].split('\0')
                  [0]+" by " + songKey[j][0].split('\0')[1])
        k = input("Please enter the number of the song you want to use: ")
        songKey = songKey[int(k)-1][0]
    else:
        songKey = songKey[0][0]
    print("Found Song: " + songKey.split('\0')
          [0]+" by " + songKey.split('\0')[1])

    root = os.path.abspath(os.getcwd()) + r'\tmp'
    pathList = []
    for path, subdirs, files in os.walk(root):
        for name in files:
            if not name.startswith("namelist"):
                pathList.append(os.path.join(path, name))

    songValue = None
    # Find song value in the database
    for path in pathList:
        with open(path, 'rb') as handle:
            data = pickle.load(handle)
            handle.close()
            if songKey in data.keys():
                songValue = data[songKey]
                break
        del data
        gc.collect()
    if len(pathList) > 0:
        del data
        gc.collect()
    if songValue == None:
        pass  # ! Error: Song not found

    cProfile.runctx('main(songValue, pathList)', {'songValue': songValue, 'pathList': pathList, 'main': main}, {}, sort='tottime', filename='profile.txt')
    p = pstats.Stats('profile.txt')
    p.sort_stats(SortKey.TIME, SortKey.CUMULATIVE).print_stats(100)

def printValues(values):
    '''
    Prints the values of the song from the database.
    '''
    print('Song: ' + values[0])
    print('Band: ' + values[1])
    print('Duration: ' + str(values[2]))
    print('Key: ' + str(values[3]))
    print('Mode: ' + str(values[4]))
    print('Speed: ' + str(values[5]))
    print('Loudness: ' + str(values[6]))
    print('Time signature: ' + str(values[7]))
    print('Year: ' + str(values[8]))
    print()

def compareValues(values1, values2):
    '''
    Prints the results of FeatureSimilarity.
    '''
    print("Same artist: " + str(FeatureSimilarity.methodDictionary[1](values1[1], values2[1])))
    print("Duration similarity: " + str(FeatureSimilarity.methodDictionary[2](values1[2], values2[2])))
    print("Key similarity: " + str(FeatureSimilarity.methodDictionary[3](values1[3], values2[3], values1[4], values2[4])))
    print("Speed similarity: " + str(FeatureSimilarity.methodDictionary[5](values1[5], values2[5], values1[7], values2[7])))
    print("Loudness similarity: " + str(FeatureSimilarity.methodDictionary[6](values1[6], values2[6])))
    print("Year similarity: " + str(FeatureSimilarity.methodDictionary[8](values1[8], values2[8])))
    print("Sections start similarity: " + str(FeatureSimilarity.methodDictionary[9](values1[9], values2[9])))
    print("Segments pitches similarity: " + str(FeatureSimilarity.methodDictionary[10](values1[10], values2[10])))
    print("Segments timbre similarity: " + str(FeatureSimilarity.methodDictionary[11](values1[11], values2[11])))
    print("Bars start similarity: " + str(FeatureSimilarity.methodDictionary[12](values1[12], values2[12])))
    print("Beats start similarity: " + str(FeatureSimilarity.methodDictionary[13](values1[13], values2[13])))
    print("Tatums start similarity: " + str(FeatureSimilarity.methodDictionary[14](values1[14], values2[14])))
    print()
    print("Overall similarity: " + str(Search.cosineSimilarity(values1, values2)))
    print()



def compareSongs(values1, values2):
    '''
    Compares the values of two songs in the database.
    '''

    plt.figure()

    #subplot(r,c) provide the no. of rows and columns
    f, axarr = plt.subplots(1, 4) 

    # normalize 2d array to 0-1
    a = (np.array(values1[10])-np.ones((len(values1[10]), len(values1[10][0])))*np.min(values1[10])) / (np.max(values1[10])-np.min(values1[10]))
    axarr[0].imshow(np.array([[(j, 0, 0) for j in i] for i in a]), interpolation='nearest')
    
    a = (np.array(values1[11])-np.ones((len(values1[11]), len(values1[11][0])))*np.min(values1[11])) / (np.max(values1[11])-np.min(values1[11]))
    axarr[1].imshow(np.array([[(j, 0, 0) for j in i] for i in a]), interpolation='nearest')

    a = (np.array(values2[10])-np.ones((len(values2[10]), len(values2[10][0])))*np.min(values2[10])) / (np.max(values2[10])-np.min(values2[10]))
    axarr[2].imshow(np.array([[(j, 0, 0) for j in i] for i in a]), interpolation='nearest')

    a = (np.array(values2[11])-np.ones((len(values2[11]), len(values2[11][0])))*np.min(values2[11])) / (np.max(values2[11])-np.min(values2[11]))
    axarr[3].imshow(np.array([[(j, 0, 0) for j in i] for i in a]), interpolation='nearest')

    axarr[0].title.set_text('Pitches 1')
    axarr[1].title.set_text('Timbre 1')
    axarr[2].title.set_text('Pitches 2')
    axarr[3].title.set_text('Timbre 2')

    printValues(values1)
    printValues(values2)

    compareValues(values1, values2)
    
    plt.show()

if __name__ == '__main__':
    preformanceTest()

File: SAR/C++/src/debug.cpp
#include "debug.hpp"
#include "Search.hpp"
#include "FeatureSimilarity.hpp"
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <thread>
#include <queue>
#include <random>
#include <fstream>
#include <stdexcept>
#include <ctime>
#include <cstdlib>

namespace Debug
{

    void main(const std::vector<float> &songValue, const std::vector<std::string> &pathList)
    {
        // Find most similar song using cosine similarity
        int numOfSongs = 5;
        std::vector<std::pair<std::string, float>> similarSongs = Search::multiProcessing(
            Search::findSimilarSongs, 32, songValue, pathList, numOfSongs);

        // Sort the list by the similarity score
        std::sort(similarSongs.begin(), similarSongs.end(), Search::takeSecond);
        if (similarSongs.size() > static_cast<size_t>(numOfSongs))
        {
            similarSongs.resize(numOfSongs);
        }

        std::cout << "Similar Songs: " << std::endl;
        for (const auto &song : similarSongs)
        {
            std::vector<std::string> songParts;
            split(song.first, '\0', songParts);
            std::cout << "Found Song: " << songParts[0] << " by " << songParts[1] << std::endl;
            std::cout << "Cosine Similarity: " << song.second << std::endl;
        }
    }

    void performanceTest()
    {
        // Ask for the song title
        std::string songTitle;
        std::cout << "Enter the song title: ";
        std::cin >> songTitle;

        // Find the song in the database
        std::string songKey = Search::fuzzyGetSongTitle(songTitle, std::string(os::getcwd()) + r '\tmp\namelist.pickle', 40);

        if (songKey.size() > 1)
        {
            std::cout << "Multiple songs found:" << std::endl;
            for (size_t j = 0; j < songKey.size(); j++)
            {
                std::vector<std::string> songParts;
                split(songKey[j], '\0', songParts);
                std::cout << j + 1 << ". " << songParts[0] << " by " << songParts[1] << std::endl;
            }
            std::cout << "Please enter the number of the song you want to use: ";
            int k;
            std::cin >> k;
            songKey = songKey[k - 1];
        }
        else
        {
            songKey = songKey[0];
        }

        std::vector<std::string> pathList;
        std::string root = std::string(os::getcwd()) + r '\tmp';
        for (const auto &path : os::walk(root))
        {
            for (const auto &name : path[2])
            {
                if (!startsWith(name, "namelist"))
                {
                    pathList.push_back(os::path.join(path[0], name));
                }
            }
        }

        std::vector<float> songValue;
        // Find song value in the database
        for (const auto &path : pathList)
        {
            std::ifstream file(path, std::ios::binary);
            if (file)
            {
                std::unordered_map<std::string, std::vector<float>> data;
                pickle::load(file, data);
                file.close();
                if (data.find(songKey) != data.end())
                {
                    songValue = data[songKey];
                    break;
                }
            }
        }

        if (songValue.empty())
        {
            // Error: Song not found
            return;
        }

        // Timing the main function
        std::clock_t start = std::clock();
        main(songValue, pathList);
        std::clock_t end = std::clock();
        double elapsed_time = double(end - start) / CLOCKS_PER_SEC;
        std::cout << "Time elapsed: " << elapsed_time << " seconds" << std::endl;
    }

    void printValues(const std::vector<float> &values)
    {
        std::cout << "Song: " << values[0] << std::endl;
        std::cout << "Band: " << values[1] << std::endl;
        std::cout << "Duration: " << values[2] << std::endl;
        std::cout << "Key: " << values[3] << std::endl;
        std::cout << "Mode: " << values[4] << std::endl;
        std::cout << "Speed: " << values[5] << std::endl;
        std::cout << "Loudness: " << values[6] << std::endl;
        std::cout << "Time signature: " << values[7] << std::endl;
        std::cout << "Year: " << values[8] << std::endl;
        std::cout << std::endl;
    }

    void compareValues(const std::vector<float> &values1, const std::vector<float> &values2)
    {
        std::cout << "Same artist: " << FeatureSimilarity::methodDictionary[1](values1[1], values2[1]) << std::endl;
        std::cout << "Duration similarity: " << FeatureSimilarity::methodDictionary[2](values1[2], values2[2]) << std::endl;
        std::cout << "Key similarity: " << FeatureSimilarity::methodDictionary[3](values1[3], values2[3], values1[4], values2[4]) << std::endl;
        std::cout << "Speed similarity: " << FeatureSimilarity::methodDictionary[5](values1[5], values2[5], values1[7], values2[7]) << std::endl;
        std::cout << "Loudness similarity: " << FeatureSimilarity::methodDictionary[6](values1[6], values2[6]) << std::endl;
        std::cout << "Year similarity: " << FeatureSimilarity::methodDictionary[8](values1[8], values2[8]) << std::endl;
        std::cout << "Sections start similarity: " << FeatureSimilarity::methodDictionary[9](values1[9], values2[9]) << std::endl;
        std::cout << "Segments pitches similarity: " << FeatureSimilarity::methodDictionary[10](values1[10], values2[10]) << std::endl;
        std::cout << "Segments timbre similarity: " << FeatureSimilarity::methodDictionary[11](values1[11], values2[11]) << std::endl;
        std::cout << "Bars start similarity: " << FeatureSimilarity::methodDictionary[12](values1[12], values2[12]) << std::endl;
        std::cout << "Beats start similarity: " << FeatureSimilarity::methodDictionary[13](values1[13], values2[13]) << std::endl;
        std::cout << "Tatums start similarity: " << FeatureSimilarity::methodDictionary[14](values1[14], values2[14]) << std::endl;
        std::cout << std::endl;
        std::cout << "Overall similarity: " << Search::cosineSimilarity(values1, values2) << std::endl;
        std::cout << std::endl;
    }

    void compareSongs(const std::vector<float> &values1, const std::vector<float> &values2)
    {
        // Assuming values1[10] and values2[10] are 2D arrays (images) represented as vectors of floats
        cv::Mat image1(values1[10].size(), values1[10][0].size(), CV_32FC1);
        cv::Mat image2(values2[10].size(), values2[10][0].size(), CV_32FC1);

        // Copy data from values1[10] and values2[10] to OpenCV Mats
        for (int i = 0; i < image1.rows; ++i)
        {
            for (int j = 0; j < image1.cols; ++j)
            {
                image1.at<float>(i, j) = values1[10][i][j];
                image2.at<float>(i, j) = values2[10][i][j];
            }
        }

        // Normalize images to the range [0, 1]
        cv::normalize(image1, image1, 0, 1, cv::NORM_MINMAX);
        cv::normalize(image2, image2, 0, 1, cv::NORM_MINMAX);

        // Create windows for visualization
        cv::namedWindow("Pitches 1", cv::WINDOW_NORMAL);
        cv::namedWindow("Timbre 1", cv::WINDOW_NORMAL);
        cv::namedWindow("Pitches 2", cv::WINDOW_NORMAL);
        cv::namedWindow("Timbre 2", cv::WINDOW_NORMAL);

        // Display images
        cv::imshow("Pitches 1", image1);
        cv::imshow("Timbre 1", image2);
        cv::imshow("Pitches 2", image1);
        cv::imshow("Timbre 2", image2);

        // Wait for user input to close windows
        cv::waitKey(0);

        // Destroy windows
        cv::destroyAllWindows();
    }

} // namespace Debug
